src/pru-motion-queue_test.cc

```diff
@@ -44,188 +44,122 @@ class MockPRUInterface : public PruHardwareInterface {
-    for (int i = 0; i < QUEUE_LEN; ++i) {
-      mmap->ring_buffer[i].state = STATE_EMPTY;
+  void SimRun(int num_exec, const uint32_t loops_left,
+              bool last_not_executed = true) {
+    // Simulate the execution of num_exec motion segments
+    for (int i=0; i < num_exec; ++i) {
+      execution_index_ = (execution_index_ + 1) % QUEUE_LEN;
+      assert(mmap->ring_buffer[execution_index_].state != STATE_EMPTY);
+      mmap->ring_buffer[execution_index_].state = STATE_EMPTY;
     }
+    if (last_not_executed || loops_left )
+      mmap->ring_buffer[execution_index_].state = STATE_FILLED;
+    mmap->status.index = execution_index_;
+    mmap->status.counter = loops_left;
   }
 
 private:
   struct MockPRUCommunication *mmap;
+  unsigned int execution_index_;
 };
 
-// Check that on init, the initial position is 0.
-TEST(RealtimePosition, init_pos) {
-  MotorsRegister absolute_pos_loops;
-  MockPRUInterface *pru_interface = new MockPRUInterface();
-  HardwareMapping *hmap = new HardwareMapping();
-  PRUMotionQueue motion_backend(hmap, (PruHardwareInterface*) pru_interface);
-
-  motion_backend.GetMotorsLoops(&absolute_pos_loops);
-
-  const MotorsRegister expected = {0, 0, 0, 0, 0, 0, 0, 0};
-  EXPECT_THAT(expected, ::testing::ContainerEq(absolute_pos_loops));
+static void check_buffer_value(PRUMotionQueue *motion_backend,
+                               const unsigned size) {
+  const unsigned int buffer = motion_backend->GetPendingElements(NULL);
+  EXPECT_EQ(buffer, size);
+}
 
-  delete pru_interface;
-  delete hmap;
+TEST(PruMotionQueue, status_init) {
+  MotorsRegister absolute_pos_loops;
+  MockPRUInterface pru_interface = MockPRUInterface();
+  HardwareMapping hmap = HardwareMapping();
+  PRUMotionQueue motion_backend(&hmap, (PruHardwareInterface*) &pru_interface);
+  check_buffer_value(&motion_backend, 0);
 }
 
-// Check that SimRun(1, 0) produce the same absolute position as
-// SimRun(2, MAX_SEGMENT_LOOPS).
-TEST(RealtimePosition, zero_loops_edge) {
-  //motor 0 +, motor 1 -, motor 2 -, 150 loops. fractions: /1 /3 /5
-  static struct MotionSegment segment = {
-    STATE_FILLED /*state*/, 0x00 | 1 << 1 | 1 << 2 /*direction bits*/,
-    0 /*loops accel*/, 150u /*loops travel*/, 0 /*loops decel*/,
-    0 /*aux*/, 0 /*accel_series_index*/, 0 /*hires_accel_cycles*/,
-    0 /*travel_delay_cycles*/,
-    { 0xFFFFFFFF, 0x55555555, 0x33333333, 0, 0, 0, 0, 0}/*fractions*/,
-  };
+TEST(PruMotionQueue, single_exec) {
   MotorsRegister absolute_pos_loops;
-  MockPRUInterface *pru_interface = new MockPRUInterface();
-  HardwareMapping *hmap = new HardwareMapping();
-  PRUMotionQueue motion_backend(hmap, (PruHardwareInterface*) pru_interface);
+  MockPRUInterface pru_interface = MockPRUInterface();
+  HardwareMapping hmap = HardwareMapping();
+  PRUMotionQueue motion_backend(&hmap, (PruHardwareInterface*) &pru_interface);
 
-  motion_backend.Enqueue(&segment);
+  static struct MotionSegment segment = {0};
```

Comment:
(Solved)

I don't really remember either, but I  cannot think for a reason right now